metals/src/main/scala/scala/meta/internal/metals/mcp/MetalsMcpServer.scala

```diff
@@ -282,15 +279,46 @@ class MetalsMcpServer(
           .compileFile(path)
           .map {
             case Some(_) =>
-              val result = diagnostics
-                .forFile(path)
-                .map { d =>
-                  val startLine = d.getRange().getStart().getLine()
-                  val endLine = d.getRange().getEnd().getLine()
-                  s"($startLine-$endLine):\n${d.getMessage()}"
+              lazy val buildTarget = buildTargets.inverseSources(path)
+
+              def inFileErrors = {
+                val errors = diagnostics.forFile(path).show()
+                if (errors.isEmpty) None else Some(s"Errors in file:\n$errors")
+              }
+
+              def inModuleErrors =
+                for {
+                  bt <- buildTarget
+                  if diagnostics.hasCompilationErrors(bt)
+                } yield {
+                  val errors =
+                    diagnostics.allDiagnostics.filter { case (path, _) =>
+                      buildTargets.inverseSources(path).contains(bt)
+                    }
+                  s"Compile errors in the module:\n${errors.show(projectPath)}"
+                }
+
+              def inUpstreamModulesErrors =
+                for {
+                  bt <- buildTarget
+                  upstreamModules = diagnostics
+                    .upstreamTargetsWithCompilationErrors(bt)
+                  if upstreamModules.nonEmpty
+                } yield {
+                  val modules = upstreamModules
+                    .flatMap(buildTargets.jvmTarget)
+                    .map(_.displayName)
+                    .mkString("\n", "\n", "")
+                  s"Compile errors in upstream modules: $modules"
                 }
-                .mkString("\n")
-              new CallToolResult(createContent(result), false)
+
+              val content = inFileErrors
```

Comment:
Shouldn't we join all of them instead of orElse?

Reply:
We can do that. I assumed the errors in current file are the most relevant and I wanted to avoid sending all other possibly unrelated errors.

Reply:
All issues from this build target? And if none, information about upstream build targets that don't compile (if upstream build target doesn't compile, we clear diagnostics in downstream to avoid stale errors)?

Reply:
Scratch that, maybe it doesn't make sense to send everything after thinking about it again. Ignore me

Reply:
But I think it may be useful to explicitly explain the behaviour in the description of the endpoint.